tools/clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -5097,14 +5097,65 @@ static void ProcessDeclAttribute(Sema &S, Scope *scope, Decl *D,
   }
 }
 
+void Sema::ValidateShaderAttribtues(Scope *S, Decl *D, const AttributeList *A) {
+  bool hasComputeAttr = false;
+  bool hasNodeAttr = false;
+  int attrNum = 0;
+
+  for (const AttributeList *l = A; l; l = l->getNext()) {
+    if (l->getKind() == AttributeList::AT_HLSLShader) {
+      Expr *E = l->getArgAsExpr(0);
+      if (E->isTypeDependent() || E->isValueDependent() ||
+          E->getStmtClass() != Stmt::StringLiteralClass) {
+        break;
+      }
+
+      StringLiteral *sl = cast<StringLiteral>(E);
+      std::string str = sl->getString();
+      if (str == "compute") {
+        hasComputeAttr = true;
+        attrNum++;
+      } else if (str == "node") {
+        hasNodeAttr = true;
+        attrNum++;
+      } else {
+        for (auto shaderAttrStr :
+             {"vertex", "pixel", "hull", "domain", "geometry", "raygeneration",
+              "intersection", "anyhit", "closesthit", "miss", "callable",
+              "mesh", "amplification"}) {
+          if (str == shaderAttrStr) {
+            attrNum++;
+            break;
+          }
+        }
```

Comment:
I think you can structure this slightly differently in a way that will both simplify it and make it more maintainable if we add other attributes. Instead of tracking `hasComputeAttr` and `hasNodeAttr`, we could keep a `shaderStage` (which is either empty or one of "compute", "vertex", etc), and `hasNodeAttr`. Then when we see a shader stage we emit the error if 1) there already is one, or 2) hasNodeAttr is true and the stage is not "compute". When we see "node" we emit an error if the shader stage is set and not "compute".